tests/util/static_check.hpp

```diff
@@ -11,20 +11,86 @@
 
 namespace util {
 
-struct static_check {
-  template <typename OperationType, typename... HandlerTypes> struct bind_left {
-    template <typename OperandType> static constexpr bool invocable(auto &&fn)
+template <typename OperandType> struct lvalue {
+  using type = std::add_lvalue_reference_t<OperandType>;
+};
+
+template <typename OperandType> struct rvalue {
+  using type = std::add_rvalue_reference_t<OperandType>;
+};
+
+template <typename OperandType> struct clvalue {
+  using type = std::add_lvalue_reference_t<std::add_const_t<OperandType>>;
+};
+
+template <typename OperandType> struct crvalue {
+  using type = std::add_rvalue_reference_t<std::add_const_t<OperandType>>;
+};
+
+template <typename OperandType> struct cvalue {
+  using type = std::add_const_t<OperandType>;
+};
+
+template <typename OperandType> struct prvalue {
+  using type = OperandType;
+};
+
+template <typename OperationType, typename OperandType> struct static_check {
+  template <typename... HandlerTypes> struct bind_right {
+    template <template <typename> typename... Categories> static constexpr auto invocable(auto &&fn) noexcept -> bool
     {
-      return fn::monadic_invocable<OperationType, OperandType, HandlerTypes..., decltype(fn)>;
+      return (fn::monadic_invocable< //
+                  OperationType, typename Categories<OperandType>::type, decltype(fn), HandlerTypes...>
+              && ...);
+    }
+
+    template <template <typename> typename... Categories>
+    static constexpr auto not_invocable(auto &&fn) noexcept -> bool
+    {
+      return (not fn::monadic_invocable< //
+                  OperationType, typename Categories<OperandType>::type, decltype(fn), HandlerTypes...>
+              && ...);
+    }
+
+    static constexpr auto invocable_with_any(auto &&fn) noexcept -> bool
+    {
+      return invocable<lvalue, cvalue, rvalue, clvalue, crvalue, prvalue>(FWD(fn));
+    }
+
+    static constexpr auto not_invocable_with_any(auto &&fn) noexcept -> bool
```

Comment:
Names for these should really be `not_invocable_with_any_value_category` but it's so long that it multilines lots of places and defeats the purpose of having this "alias" in the first place. I could not find a shorter name that still sort of made sense in this context. Please do suggest better name if you can think of one :rage3: